### ISSUE TITLE
PopSynth: PISN with companion does not get redirected correctly.

### ISSUE THREAD

**Issue by maxbriel** [POSYDON-code/POSYDON-public#195]
**Problem**
If a PISN occurs with a BH companion, the binary is not redirected to `step_end`.
Instead, it just ends and causes CC2 to appear twice in the history of the binary.
Similarly, if a PISN occurs with a stellar companion CC1 appears twice and the secondary is not evolved.

**Solution**
Redirect the remaining BH to the `end_step`.
Redirect the companion to single stars?

**Example CC1**

time | state | step_names | event | S1_mass | S2_mass | orbital_period | eccentricity | step_times
-- | -- | -- | -- | -- | -- | -- | -- | --
0.000000e+00 | detached | initial_cond | ZAMS | 175.592698 | 172.834458 | 5.317596 | 0.0 | 0.000000
2.759087e+06 | detached | step_HMS_HMS | CC1 | 132.408030 | 218.595768 | 6.317904 | 0.0 | 0.026219
2.759087e+06 | detached | step_SN | CC1 | NaN | 218.595768 | 6.317904 | 0.0 | 0.000059

**Example CC2**

time | state | step_names | event | S1_mass | S2_mass | orbital_period | eccentricity | metallicity
-- | -- | -- | -- | -- | -- | -- | -- | --
0.000000e+00 | detached | initial_cond | ZAMS | 176.446284 | 170.928690 | 69.374127 | 0.000000 | 0.00142
2.803116e+06 | detached | step_HMS_HMS | CC1 | 150.071558 | 142.702775 | 100.799688 | 0.000000 | 0.00142
2.803116e+06 | detached | step_SN | NaN | 148.168377 | 142.702775 | 102.201234 | 0.008509 | 0.00142
2.851108e+06 | detached | step_detached | CC2 | 148.168377 | 137.473264 | 105.095544 | 0.007258 | 0.00142
2.851108e+06 | detached | step_SN | CC2 | 148.168377 | NaN | 105.095544 | 0.007258 | 0.00142

**Comment by maxbriel** [comment]
These can be found in the population file: 
- BH companion + PISN: `6073`
`230914/POSYDON_data/tutorials/population-synthesis/example/1.00e-01_Zsun_population.h5`

- Stellar companion + PISN: `257601`
`230914/POSYDON_data/tutorials/population-synthesis/example/1.00e+00_Zsun_population.h5`

**Comment by maxbriel** [comment]
Similar to Issue #220, the interpolator output gives `PISN` as a state of star 1 and `CCSN` as the event (for `MODEL05`). This is an undefined state and in `step_SN` it crashed.

This does not seem like the correct output for the interpolator. This should be clarified.
If it is the correct output, then `step_SN` needs to correctly deal with it and not use `SN_type` to identify the PISN events.


Here's code to recreate the issue:

```
from posydon.binary_evol.binarystar import BinaryStar, SingleStar
from posydon.binary_evol.simulationproperties import SimulationProperties
from posydon.popsyn.io import simprop_kwargs_from_ini


sim_kwargs = simprop_kwargs_from_ini('population.ini', verbose=True)
metallicity = {'metallicity':0.01,}
sim_kwargs['step_HMS_HMS'][1].update(metallicity)
sim_kwargs['step_CO_HeMS'][1].update(metallicity)
sim_kwargs['step_CO_HMS_RLO'][1].update(metallicity)
sim_kwargs['step_CO_HeMS_RLO'][1].update(metallicity)
sim_kwargs['step_detached'][1].update(metallicity)
sim_kwargs['step_disrupted'][1].update(metallicity)
sim_kwargs['step_merged'][1].update(metallicity)
sim_kwargs['step_initially_single'][1].update(metallicity)
sim_pop = SimulationProperties(**sim_kwargs)
sim_pop.load_steps(verbose=False)

STAR1 = SingleStar(**{'mass': 	170.638207, 
                      'state': 'H-rich_Core_H_burning',
                      'natal_kick_array': [4.921294, 4.31745, 1.777768, 3.509656]})
STAR2 = SingleStar(**{'mass':37.917852,
                      'state': 'H-rich_Core_H_burning'})

BINARY = BinaryStar(STAR1, STAR2,  
                    **{'time': 0.0, 'state': 'detached', 'event': 'ZAMS', 'orbital_period':113.352736, 'eccentricity': 0.0},
                    properties = sim_pop)

try:
    BINARY.evolve()
except:
    print('We failed!')
    
BINARY.to_df()[['state', 'event', 'S1_mass', 'S2_mass', 'orbital_period', 'S1_state', 'S2_state']]
```

**Comment by ezapartas** [comment]
@maxbriel Just to add one comments here. If PISN occurs, the binary.state should become disrupted , and this way, the companion would keep evolving is step_disrupted (subclass of step_isolated, the latter based on and being a subclass of step_detached).

**Comment by maxbriel** [comment]
I'm working on a solution in PR #240.
There, I am giving priority to the CO_type interpolator.

@ezapartas With the PISN, should the remaining star become a `massless_remnant` as used in `step_merged`?

With the mass set to 0.0 and the state to `massless_remnant`, the calculation of `orbit_kick` should always lead to a `step_disrupted`. 
However, this leads for the system to be tagged as a `disrupted Thorne-Zytkow object`, when the second star becomes a compact object. Should the logic in `step_detached` be adjusted for this? 

<details><summary>Binary evolution details</summary>
<p>

state | event | time | S1_mass | S2_mass | orbital_period | S1_state | S2_state | step_names
-- | -- | -- | -- | -- | -- | -- | -- | --
detached | ZAMS | 0.000000e+00 | 170.638207 | 37.917852 | 113.352736 | H-rich_Core_H_burning | H-rich_Core_H_burning | initial_cond
detached | CC1 | 2.827616e+06 | 155.652108 | 38.967189 | 71.996420 | H-rich_Central_He_depleted | H-rich_Core_H_burning | step_HMS_HMS
disrupted | <NA> | 2.827616e+06 | 0.000000 | 38.967189 | NaN | massless_remnant | H-rich_Core_H_burning | step_SN
disrupted | CC2 | 5.799377e+06 | 0.000000 | 29.299084 | NaN | massless_remnant | H-rich_Central_C_depletion | step_disrupted
disrupted | <NA> | 5.799377e+06 | 0.000000 | 19.314849 | NaN | massless_remnant | BH | step_SN
disrupted Thorne-Zytkow object | <NA> | 5.799377e+06 | 0.000000 | 19.314849 | NaN | massless_remnant | BH | step_disrupted
disrupted Thorne-Zytkow object | END | 5.799377e+06 | 0.000000 | 19.314849 | NaN | massless_remnant | BH | nan


</p>
</details> 

When just setting the mass to 0.0 and the state=None, the binary immediately goes to `END` and the secondary is not properly evolved. 

<details><summary>mass=0.0 option</summary>
<p>


   state | event | time | S1_mass | S2_mass | orbital_period | S1_state | S2_state | step_names
 -- | -- | -- | -- | -- | -- | -- | -- | --
detached | ZAMS | 0.000000e+00 | 170.638207 | 37.917852 | 113.352736 | H-rich_Core_H_burning | H-rich_Core_H_burning | initial_cond
detached | CC1 | 2.827616e+06 | 155.652108 | 38.967189 | 71.996420 | H-rich_Central_He_depleted | H-rich_Core_H_burning | step_HMS_HMS
disrupted | <NA> | 2.827616e+06 | 0.000000 | 38.967189 | NaN | <NA> | H-rich_Core_H_burning | step_SN
disrupted | END | 2.827616e+06 | 0.000000 | 38.967189 | NaN | <NA> | H-rich_Core_H_burning | nan



</p>
</details>

**Comment by ezapartas** [comment]
The change to the `massless_remnant` is correct , and indeed the step_disrupted is called correctly. Indeed the issue is in the  step_detached, lines ~1020 and ~1038 where, in a hard coded way, assumed that a massless + CO in detached is a Thorne-Zytkov merger (thinking that massless remnants is the outcome of a merger (or having an initially single stars), we did not think of a massless remnant being the outcome of PISN).

**Comment by ezapartas** [comment]
In my opinion, we can even comment the TZ flag and just not add anything if massless+CO. People that are interesting in TZ mergers (i.e. star + CO) to parse and analyze the post-processingly (the same way that we do not flag GRBs in POSYDON, only in the parse and synthetic_popoulation analysis).

**Comment by maxbriel** [comment]
I agree with you and I've adapted the `step_detached` accordingly. This actually required some additional changes in the flowchart to send a disrupted binary with a massless remnant and CO to `step_end`. (there was also a line missing there)

Because of these changes, I will check if a previously TZO object with state `('massless_remnant', {CO}, 'merged', {ANY_BINARY_EVENT})` will still go to the correct final step.

**Comment by maxbriel** [comment]
Solved with PR #240
